noir/noir-repo/aztec_macros/src/transforms/note_interface.rs

```diff
@@ -0,0 +1,510 @@
+        //     }
+        // }
+        // It assumes every note field is stored in an individual serialized field,
+        // and can be converted to the original type via the from_field() trait (structs) or cast as Field (integers)
+
+        if !check_trait_method_implemented(trait_impl, "deserialize_content") {
+            let note_deserialize_content_fn = generate_note_deserialize_content(
+                &note_type,
+                &note_fields,
+                &note_serialized_len,
+                note_interface_impl_span,
+            )?;
+            trait_impl.items.push(TraitImplItem::Function(note_deserialize_content_fn));
+        }
+
+        // Automatically generate a struct that represents the note's serialization metadata, and an auxiliary function to retrieve it
+        // The struct looks like this
+        //
+        // NoteTypeFields {
+        //     field1: FieldSelector { index: 0, offset: 0, length: 32 },
+        //     field2: FieldSelector { index: 1, offset: 0, length: 32 },
+        //     ...
+        // }
+        //
+        // It assumes each field occupies an entire field and its serialized in definition order
+
+        if !note_impl.methods.iter().any(|(func, _)| func.def.name.0.contents == "fields") {
+            let note_fields_struct =
+                generate_note_fields_struct(&note_type, &note_fields, note_interface_impl_span)?;
+            note_fields_structs.push(note_fields_struct);
+            let note_fields_fn =
+                generate_note_fields_fn(&note_type, &note_fields, note_interface_impl_span)?;
+            note_impl.methods.push((note_fields_fn, note_impl.type_span));
+        }
+
+        // Automatically generate the header getter and setter methods
+        if !check_trait_method_implemented(trait_impl, "get_header") {
+            let get_header_fn = generate_note_get_header(&note_type, note_interface_impl_span)?;
+            trait_impl.items.push(TraitImplItem::Function(get_header_fn));
+        }
+        if !check_trait_method_implemented(trait_impl, "set_header") {
+            let set_header_fn = generate_note_set_header(&note_type, note_interface_impl_span)?;
+            trait_impl.items.push(TraitImplItem::Function(set_header_fn));
+        }
+
+        // Automatically generate the note type id getter method. The id itself its calculated as the concatenation
+        // of the conversion of the characters in the note's struct name to unsigned integers.
+        if !check_trait_method_implemented(trait_impl, "get_note_type_id") {
+            let get_note_type_id_fn =
+                generate_note_get_type_id(&note_type, note_interface_impl_span)?;
+            trait_impl.items.push(TraitImplItem::Function(get_note_type_id_fn));
+        }
+
+        // Automatically generate the method to compute the note's content hash as:
+        // fn compute_note_content_hash(self: NoteType) -> Field {
+        //    // TODO(#1205) Should use a non-zero generator index.
```

Comment:
Love that we generate invisible code with TODOs in it :-P

Reply:
Kept it so a generic search of TODO(#1205) would yield the result in the codebase, but I know it's weird 🤣